sycl/source/detail/scheduler/commands.cpp

```diff
@@ -1743,7 +1728,36 @@ pi_result ExecCGCommand::SetKernelParamsAndLaunch(
       break;
     }
     }
-    ++NextTrueIndex;
+  };
+
+  if (EliminatedArgMask.empty()) {
+    for (ArgDesc &Arg : Args) {
+      setFunc(Arg, Arg.MIndex);
+    }
+  } else {
+    // TODO this is not necessary as long as we can guarantee that the arguments
+    // are already sorted (e. g. handle the sorting in handler if necessary due
+    // to set_arg(...) usage).
+    std::sort(Args.begin(), Args.end(), [](const ArgDesc &A, const ArgDesc &B) {
+      return A.MIndex < B.MIndex;
+    });
+    int LastIndex = -1;
+    size_t NextTrueIndex = 0;
+
+    for (ArgDesc &Arg : Args) {
+      // Handle potential gaps in set arguments (e. g. if some of them are set
+      // on the user side).
+      for (int Idx = LastIndex + 1; Idx < Arg.MIndex; ++Idx)
+        if (EliminatedArgMask.empty() || !EliminatedArgMask[Idx])
```

Comment:
EliminatedArgsMask.empty() is always false here because it is checked at the line 1733
```suggestion
        if (!EliminatedArgMask[Idx])
```
Please also do similar change below at the line 1755